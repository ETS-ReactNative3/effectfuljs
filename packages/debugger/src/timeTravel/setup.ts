import config from "../config";
config.timeTravel = true;
config.mutationObserver = true;
