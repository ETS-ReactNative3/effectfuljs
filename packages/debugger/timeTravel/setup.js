"use strict";

var _config = require("../config");

_config.default.timeTravel = true;
_config.default.mutationObserver = true;