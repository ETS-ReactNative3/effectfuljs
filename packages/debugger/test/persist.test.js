require("./setup");
const trace = require("./run").bind(null, true);
const D = require("../main");
const S = require("@effectful/serialization");

test("persistent state", function() {
  global.console = { log: jest.fn() };
  D.evalThunk(require("./__fixtures__/counters"));
  const extra = { value: "H" };
  D.Persist.extraState().add(extra);
  const mod = trace();
  S.regOpaqueObject(mod, "countersMod");
  mod.runCounters(true);
  const [counter1, s1] = trace();
  console.log(counter1, s1);
  D.Persist.restore(s1);
  const [extraAfter] = D.Persist.extraState();
  expect(extra).not.toBe(extraAfter);
  expect(extra).toEqual(extraAfter);
  const [counter2, s2] = trace();
  console.log(counter2, s2);
  // expect(s2).toEqual(s1);
  expect(console.log.mock.calls).toMatchSnapshot();
});
