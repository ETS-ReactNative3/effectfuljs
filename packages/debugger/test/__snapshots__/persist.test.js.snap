// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`persistent state 1`] = `
Array [
  Array [
    0,
    Array [
      Object {
        "i": 0,
        "u": 4,
      },
    ],
    Array [
      Array [
        Object {
          "i": 0,
          "u": 4,
        },
        0,
      ],
    ],
    Array [
      0,
    ],
  ],
  Array [
    1,
    Array [
      Object {
        "i": 0,
        "u": 4,
      },
      Object {
        "i": 1,
        "u": 14,
      },
    ],
    Array [
      Array [
        Object {
          "i": 0,
          "u": 4,
        },
        0,
      ],
      Array [
        Object {
          "i": 1,
          "u": 14,
        },
        1,
      ],
    ],
    Array [
      0,
      1,
    ],
  ],
  Array [
    2,
    Array [
      Object {
        "i": 0,
        "u": 4,
      },
      Object {
        "i": 1,
        "u": 14,
      },
      Object {
        "i": 2,
        "u": 24,
      },
    ],
    Array [
      Array [
        Object {
          "i": 0,
          "u": 4,
        },
        0,
      ],
      Array [
        Object {
          "i": 1,
          "u": 14,
        },
        1,
      ],
      Array [
        Object {
          "i": 2,
          "u": 24,
        },
        2,
      ],
    ],
    Array [
      0,
      1,
      2,
    ],
  ],
  Array [
    3,
    Array [
      Object {
        "i": 0,
        "u": 4,
      },
      Object {
        "i": 1,
        "u": 14,
      },
      Object {
        "i": 2,
        "u": 24,
      },
      Object {
        "i": 3,
        "u": 34,
      },
    ],
    Array [
      Array [
        Object {
          "i": 0,
          "u": 4,
        },
        0,
      ],
      Array [
        Object {
          "i": 1,
          "u": 14,
        },
        1,
      ],
      Array [
        Object {
          "i": 2,
          "u": 24,
        },
        2,
      ],
      Array [
        Object {
          "i": 3,
          "u": 34,
        },
        3,
      ],
    ],
    Array [
      0,
      1,
      2,
      3,
    ],
  ],
  Array [
    4,
    Object {
      "f": Array [
        Array [
          "top",
          Object {
            "r": 22,
          },
        ],
        Array [
          "syncStack",
          Array [],
        ],
        Array [
          "queue",
          Array [],
        ],
        Array [
          "debug",
          true,
        ],
        Array [
          "brk",
          Object {
            "r": 21,
          },
        ],
        Array [
          "extra",
          Object {
            "$": "Set",
            "l": Array [
              Object {
                "f": Array [
                  Array [
                    "value",
                    "H",
                  ],
                ],
              },
            ],
          },
        ],
      ],
      "x": Array [
        Object {
          "$": "i#/test/__fixtures__/counters.js#f",
        },
        Object {
          "$": "@effectful/debugger/call",
        },
        Object {
          "$": "@effectful/debugger/apply",
        },
        Object {
          "$": "p#/test/__fixtures__/counters.js#f_1",
          "$$": null,
          "f": Array [
            Array [
              "prototype",
              Object {
                "f": Array [
                  Array [
                    "constructor",
                    Object {
                      "r": 3,
                    },
                    2,
                  ],
                ],
              },
              3,
            ],
            Array [
              "call",
              Object {
                "r": 1,
              },
            ],
            Array [
              "apply",
              Object {
                "r": 2,
              },
            ],
          ],
        },
        Object {
          "$": "countersMod",
        },
        Object {
          "f": Array [
            Array [
              "exports",
              Object {
                "r": 4,
              },
            ],
          ],
        },
        Object {
          "$": "p#/test/__fixtures__/counters.js#getCounter_1",
          "$$": Object {
            "r": 12,
          },
          "f": Array [
            Array [
              "prototype",
              Object {
                "f": Array [
                  Array [
                    "constructor",
                    Object {
                      "r": 6,
                    },
                    2,
                  ],
                ],
              },
              3,
            ],
            Array [
              "call",
              Object {
                "r": 1,
              },
            ],
            Array [
              "apply",
              Object {
                "r": 2,
              },
            ],
          ],
        },
        Object {
          "$": "p#/test/__fixtures__/counters.js#runCounters_1",
          "$$": Object {
            "r": 12,
          },
          "f": Array [
            Array [
              "prototype",
              Object {
                "f": Array [
                  Array [
                    "constructor",
                    Object {
                      "r": 7,
                    },
                    2,
                  ],
                ],
              },
              3,
            ],
            Array [
              "call",
              Object {
                "r": 1,
              },
            ],
            Array [
              "apply",
              Object {
                "r": 2,
              },
            ],
          ],
        },
        Object {
          "$": "p#/src/instr/rt.ts#wrapExport_1",
          "$$": null,
          "f": Array [
            Array [
              "prototype",
              Object {
                "f": Array [
                  Array [
                    "constructor",
                    Object {
                      "r": 8,
                    },
                    2,
                  ],
                ],
              },
              3,
            ],
            Array [
              "call",
              Object {
                "r": 1,
              },
            ],
            Array [
              "apply",
              Object {
                "r": 2,
              },
            ],
          ],
        },
        Object {
          "$": "p#/src/instr/rt.ts#exportsThunk_1",
          "$$": Object {
            "r": 11,
          },
          "f": Array [
            Array [
              "prototype",
              Object {
                "f": Array [
                  Array [
                    "constructor",
                    Object {
                      "r": 9,
                    },
                    2,
                  ],
                ],
              },
              3,
            ],
            Array [
              "call",
              Object {
                "r": 1,
              },
            ],
            Array [
              "apply",
              Object {
                "r": 2,
              },
            ],
            Array [
              Object {
                "$": "@effectful/debugger/state#exports#@effectful/debugger/state#exports#thunkSymbol#Symbol(@effectful/debugger/thunk)",
              },
              true,
            ],
            Array [
              Object {
                "$": "@effectful/debugger/state#exports#@effectful/debugger/state#exports#moduleSymbol#Symbol(@effectful/debugger/module)",
              },
              Object {
                "$": "undefined",
              },
            ],
          ],
        },
        Object {
          "$": "@effectful/debugger/engine#exports#token",
        },
        Object {
          "f": Array [
            Array [
              "$$",
              null,
            ],
            Array [
              "state",
              0,
            ],
            Array [
              "goto",
              0,
            ],
            Array [
              "meta",
              Object {
                "$": "i#/src/instr/rt.ts#*_17",
              },
            ],
            Array [
              "func",
              Object {
                "r": 8,
              },
            ],
            Array [
              "self",
              Object {
                "$": "@effectful/debugger/instr/rt#exports",
              },
            ],
            Array [
              "newTarget",
              false,
            ],
            Array [
              "$",
              Object {
                "f": Array [
                  Array [
                    "top",
                    Object {
                      "r": 3,
                    },
                  ],
                  Array [
                    "descr",
                    Object {
                      "$": "undefined",
                    },
                  ],
                  Array [
                    "mod",
                    Object {
                      "r": 5,
                    },
                  ],
                  Array [
                    "res",
                    Object {
                      "r": 9,
                    },
                  ],
                ],
              },
            ],
            Array [
              "brk",
              null,
            ],
            Array [
              "next",
              null,
            ],
            Array [
              "restoreDebug",
              false,
            ],
            Array [
              "awaiting",
              Object {
                "r": 10,
              },
            ],
            Array [
              "onResolve",
              null,
            ],
            Array [
              "onReject",
              null,
            ],
            Array [
              "promise",
              null,
            ],
            Array [
              "timestamp",
              null,
            ],
          ],
        },
        Object {
          "f": Array [
            Array [
              "$$",
              null,
            ],
            Array [
              "state",
              0,
            ],
            Array [
              "goto",
              0,
            ],
            Array [
              "meta",
              Object {
                "r": 0,
              },
            ],
            Array [
              "func",
              Object {
                "r": 3,
              },
            ],
            Array [
              "self",
              Object {
                "$": "undefined",
              },
            ],
            Array [
              "newTarget",
              false,
            ],
            Array [
              "$",
              Object {
                "f": Array [
                  Array [
                    "module",
                    Object {
                      "r": 5,
                    },
                  ],
                  Array [
                    "exports",
                    Object {
                      "r": 4,
                    },
                  ],
                  Array [
                    "getCounter",
                    Object {
                      "r": 6,
                    },
                  ],
                  Array [
                    "runCounters",
                    Object {
                      "r": 7,
                    },
                  ],
                  Array [
                    "D",
                    Object {
                      "$": "../../main#exports",
                    },
                  ],
                ],
              },
            ],
            Array [
              "brk",
              Object {
                "f": Array [
                  Array [
                    "type",
                    3,
                  ],
                  Array [
                    "reason",
                    "exit",
                  ],
                  Array [
                    "id",
                    1,
                  ],
                  Array [
                    "meta",
                    Object {
                      "r": 0,
                    },
                  ],
                  Array [
                    "scope",
                    Object {},
                  ],
                  Array [
                    "line",
                    35,
                  ],
                  Array [
                    "column",
                    0,
                  ],
                  Array [
                    "endLine",
                    35,
                  ],
                  Array [
                    "endColumn",
                    0,
                  ],
                  Array [
                    "exit",
                    true,
                  ],
                  Array [
                    "location",
                    "1:?-35:?",
                  ],
                ],
              },
            ],
            Array [
              "next",
              Object {
                "f": Array [
                  Array [
                    "$$",
                    Object {
                      "r": 11,
                    },
                  ],
                  Array [
                    "state",
                    0,
                  ],
                  Array [
                    "goto",
                    0,
                  ],
                  Array [
                    "meta",
                    Object {
                      "$": "i#/src/instr/rt.ts#*_18",
                    },
                  ],
                  Array [
                    "func",
                    Object {
                      "r": 9,
                    },
                  ],
                  Array [
                    "self",
                    Object {
                      "$": "undefined",
                    },
                  ],
                  Array [
                    "newTarget",
                    false,
                  ],
                  Array [
                    "$",
                    Object {},
                  ],
                  Array [
                    "brk",
                    null,
                  ],
                  Array [
                    "next",
                    null,
                  ],
                  Array [
                    "restoreDebug",
                    false,
                  ],
                  Array [
                    "awaiting",
                    Object {
                      "r": 10,
                    },
                  ],
                  Array [
                    "onResolve",
                    null,
                  ],
                  Array [
                    "onReject",
                    null,
                  ],
                  Array [
                    "promise",
                    null,
                  ],
                  Array [
                    "timestamp",
                    null,
                  ],
                ],
              },
            ],
            Array [
              "restoreDebug",
              false,
            ],
            Array [
              "awaiting",
              Object {
                "r": 10,
              },
            ],
            Array [
              "onResolve",
              null,
            ],
            Array [
              "onReject",
              null,
            ],
            Array [
              "promise",
              null,
            ],
            Array [
              "timestamp",
              null,
            ],
          ],
        },
        Object {
          "$": "p#/test/__fixtures__/counters.js#invoke_1",
          "$$": Object {
            "r": 22,
          },
          "f": Array [
            Array [
              "prototype",
              Object {
                "f": Array [
                  Array [
                    "constructor",
                    Object {
                      "r": 13,
                    },
                    2,
                  ],
                ],
              },
              3,
            ],
            Array [
              "call",
              Object {
                "r": 1,
              },
            ],
            Array [
              "apply",
              Object {
                "r": 2,
              },
            ],
          ],
        },
        Object {
          "$": "i#/test/__fixtures__/counters.js#getCounter",
        },
        Object {
          "f": Array [
            Array [
              "i",
              0,
            ],
            Array [
              "u",
              1,
            ],
          ],
        },
        Object {
          "f": Array [
            Array [
              "i",
              1,
            ],
            Array [
              "u",
              11,
            ],
          ],
        },
        Array [
          Object {
            "r": 15,
          },
          Object {
            "r": 16,
          },
        ],
        Object {
          "$": "Map",
          "k": Array [
            Object {
              "r": 15,
            },
            Object {
              "r": 16,
            },
          ],
          "v": Array [
            0,
            1,
          ],
        },
        Object {
          "$": "Set",
          "l": Array [
            0,
            1,
          ],
        },
        Object {
          "$": "p#/test/__fixtures__/counters.js#counter_1",
          "$$": Object {
            "f": Array [
              Array [
                "$$",
                Object {
                  "r": 12,
                },
              ],
              Array [
                "state",
                0,
              ],
              Array [
                "goto",
                0,
              ],
              Array [
                "meta",
                Object {
                  "r": 14,
                },
              ],
              Array [
                "func",
                Object {
                  "r": 6,
                },
              ],
              Array [
                "self",
                Object {
                  "$": "undefined",
                },
              ],
              Array [
                "newTarget",
                false,
              ],
              Array [
                "$",
                Object {
                  "f": Array [
                    Array [
                      "i",
                      2,
                    ],
                    Array [
                      "trace",
                      Object {
                        "r": 17,
                      },
                    ],
                    Array [
                      "dir",
                      Object {
                        "r": 18,
                      },
                    ],
                    Array [
                      "set",
                      Object {
                        "r": 19,
                      },
                    ],
                  ],
                },
              ],
              Array [
                "brk",
                Object {
                  "f": Array [
                    Array [
                      "type",
                      3,
                    ],
                    Array [
                      "reason",
                      "exit",
                    ],
                    Array [
                      "id",
                      5,
                    ],
                    Array [
                      "meta",
                      Object {
                        "r": 14,
                      },
                    ],
                    Array [
                      "scope",
                      Object {
                        "f": Array [
                          Array [
                            "getCounter",
                            Array [
                              "getCounter",
                              1,
                            ],
                          ],
                          Array [
                            "runCounters",
                            Array [
                              "runCounters",
                              1,
                            ],
                          ],
                          Array [
                            "i",
                            Array [
                              "i",
                              0,
                            ],
                          ],
                          Array [
                            "trace",
                            Array [
                              "trace",
                              0,
                            ],
                          ],
                          Array [
                            "dir",
                            Array [
                              "dir",
                              0,
                            ],
                          ],
                          Array [
                            "set",
                            Array [
                              "set",
                              0,
                            ],
                          ],
                        ],
                      },
                    ],
                    Array [
                      "line",
                      21,
                    ],
                    Array [
                      "column",
                      1,
                    ],
                    Array [
                      "endLine",
                      21,
                    ],
                    Array [
                      "endColumn",
                      1,
                    ],
                    Array [
                      "exit",
                      true,
                    ],
                    Array [
                      "location",
                      "3:7-21:1",
                    ],
                  ],
                },
              ],
              Array [
                "next",
                Object {
                  "r": 22,
                },
              ],
              Array [
                "restoreDebug",
                false,
              ],
              Array [
                "awaiting",
                Object {
                  "r": 10,
                },
              ],
              Array [
                "onResolve",
                null,
              ],
              Array [
                "onReject",
                null,
              ],
              Array [
                "promise",
                null,
              ],
              Array [
                "timestamp",
                null,
              ],
            ],
          },
          "f": Array [
            Array [
              "prototype",
              Object {
                "f": Array [
                  Array [
                    "constructor",
                    Object {
                      "r": 20,
                    },
                    2,
                  ],
                ],
              },
              3,
            ],
            Array [
              "call",
              Object {
                "r": 1,
              },
            ],
            Array [
              "apply",
              Object {
                "r": 2,
              },
            ],
          ],
        },
        Object {
          "$": "s#/test/__fixtures__/counters.js#runCounters_1#3",
        },
        Object {
          "f": Array [
            Array [
              "$$",
              Object {
                "r": 12,
              },
            ],
            Array [
              "state",
              11,
            ],
            Array [
              "goto",
              12,
            ],
            Array [
              "meta",
              Object {
                "$": "i#/test/__fixtures__/counters.js#runCounters",
              },
            ],
            Array [
              "func",
              Object {
                "r": 7,
              },
            ],
            Array [
              "self",
              Object {
                "r": 4,
              },
            ],
            Array [
              "newTarget",
              false,
            ],
            Array [
              "$",
              Object {
                "f": Array [
                  Array [
                    "persist",
                    true,
                  ],
                  Array [
                    "invoke",
                    Object {
                      "r": 13,
                    },
                  ],
                  Array [
                    "c",
                    Object {
                      "r": 20,
                    },
                  ],
                  Array [
                    "trace",
                    Object {
                      "r": 17,
                    },
                  ],
                  Array [
                    "dir",
                    Object {
                      "r": 18,
                    },
                  ],
                  Array [
                    "set",
                    Object {
                      "r": 19,
                    },
                  ],
                ],
              },
            ],
            Array [
              "brk",
              Object {
                "r": 21,
              },
            ],
            Array [
              "next",
              null,
            ],
            Array [
              "restoreDebug",
              false,
            ],
            Array [
              "awaiting",
              Object {
                "r": 10,
              },
            ],
            Array [
              "onResolve",
              null,
            ],
            Array [
              "onReject",
              null,
            ],
            Array [
              "promise",
              null,
            ],
            Array [
              "timestamp",
              null,
            ],
          ],
        },
      ],
    },
  ],
  Array [
    2,
    Array [
      Object {
        "i": 0,
        "u": 4,
      },
      Object {
        "i": 1,
        "u": 14,
      },
      Object {
        "i": 2,
        "u": 24,
      },
    ],
    Array [
      Array [
        Object {
          "i": 0,
          "u": 4,
        },
        0,
      ],
      Array [
        Object {
          "i": 1,
          "u": 14,
        },
        1,
      ],
      Array [
        Object {
          "i": 2,
          "u": 24,
        },
        2,
      ],
    ],
    Array [
      0,
      1,
      2,
    ],
  ],
  Array [
    3,
    Array [
      Object {
        "i": 0,
        "u": 4,
      },
      Object {
        "i": 1,
        "u": 14,
      },
      Object {
        "i": 2,
        "u": 24,
      },
      Object {
        "i": 3,
        "u": 34,
      },
    ],
    Array [
      Array [
        Object {
          "i": 0,
          "u": 4,
        },
        0,
      ],
      Array [
        Object {
          "i": 1,
          "u": 14,
        },
        1,
      ],
      Array [
        Object {
          "i": 2,
          "u": 24,
        },
        2,
      ],
      Array [
        Object {
          "i": 3,
          "u": 34,
        },
        3,
      ],
    ],
    Array [
      0,
      1,
      2,
      3,
    ],
  ],
  Array [
    4,
    Object {
      "s2": true,
    },
  ],
]
`;
